### DISCUSSION TITLE
Is it possible define a NetworkBehaviour with fields from the runtime?

### DISCUSSION THREAD

**Discussion by scholz-timo** [libp2p/rust-libp2p#4139] [Q&A]
Hello,
I am trying to create a binding of rust-libp2p to swift, java and wasm, since the swift-libp2p and java-libp2p Projects are still unstable.

What I've done so far:
1. Created a basic wrapper project
2. Created binding for every output in different projects (To prevent macros for different outputs).
3. Created other utility files for the consumption of the given library.
4. Write wrapper?

I am stuck since I am not sure on how to implement custom routes in runtime.

Can you provide me some support so I can continue developing?

**Comment by thomaseizinger** [comment]
Can you share the code that you've got?

With anything FFI, you are likely better off keeping the API as small as possible which typically means write as much code as you can in Rust.

I'd pick a runtime (i.e. `tokio`), write the event loop in Rust and design your FFI around passing messages back and forth between the event loop and your host program.

**Reply by scholz-timo** [reply]
I initially started writing the code with swift-libp2p and java-p2p but since these are not far enough in development, I decided to switch to rust.

https://github.com/scholz-timo/capacitor-libp2p

This is the thing I want to develop.

I could use some help with adding the routes dynamically.

**Reply by thomaseizinger** [reply]
Can you point me to anything specific that you need help with? What do you mean with routes? Have you had a look at our [examples](https://github.com/libp2p/rust-libp2p/tree/master/examples) already? Do you have any experience in Rust FFIs already?

**Reply by scholz-timo** [reply]
I had a look at the examples.

I don't have that much experience with rust or rust-ffis, but I have someone on my team that does.
The problem is that we want to add:

/example:0.0.0

But not statically using the NetworkBehaviour, but rather programmatically.

So that we can then add the routes using the ffi.

**Reply by scholz-timo** [reply]
Otherwise would it be possible to schedule a call?

**Reply by mxinden** [reply]
Want to join our open maintainers call tomorrow @scholz-timo?

https://github.com/libp2p/rust-libp2p/discussions/4149

**Reply by scholz-timo** [reply]
Sure, ill leave this open to later propose a solution.

**Comment by canewsin** [comment]
@scholz-timo I tried to do it previously, you can find it here 
https://github.com/decentnetwork/decentnet
these are for dart lang though but concepts are similar.

**Comment by scholz-timo** [comment]
@thomaseizinger @mxinden I have found at least one thing that is not properly documented.

How do we get a peer id from a multiaddr?

**Reply by scholz-timo** [reply]
After using dial(...), is it only possible using the events, since it is stored in the pool it would be kind of bad to store it twice.

**Reply by thomaseizinger** [reply]
> @thomaseizinger @mxinden I have found at least one thing that is not properly documented.
> 
> How do we get a peer id from a multiaddr?

You check if the last `Protocol` is `P2p` :)

**Reply by thomaseizinger** [reply]
> After using dial(...), is it only possible using the events, since it is stored in the pool it would be kind of bad to store it twice.

Can you expand on what you mean?

**Reply by scholz-timo** [reply]
So next question.

When you want to share the swarm between multiple threads, you need to have an async mutex around it.
Since we call the next function in the async scope, is it possible to let that function timeout and return none to release the lock?

Or is this not intentended, we are not sure if we should race this future, since it contains the event and the data within might be dropped.

**Reply by scholz-timo** [reply]
Already found a solution. It is possible to call .timeout() on the ExtStream element.

**Reply by mxinden** [reply]
> When you want to share the swarm between multiple threads

I recommend not sharing the `Swarm` between threads. Instead, have a single _Task_ own the `Swarm` and communicate with the task via a `futures::mpsc::channel`.

See file sharing example.

**Reply by thomaseizinger** [reply]
To answer the question: `Swarm::next_event` is cancellation save.
